ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -0,0 +1,140 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , toBase32
+  , toBase256
+  , segwitEncode
+  , segwitDecode
+  , Word5()
+  , word5
+  , fromWord5
+  ) where
+
+import Control.Monad (guard)
+import qualified Data.Array as Arr
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor, testBit)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import Data.Ix (Ix(..))
+import Data.Word (Word8, Word32)
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+newtype Word5 = UnsafeWord5 Word8
+              deriving (Eq, Ord)
+
+instance Ix Word5 where
+  range (UnsafeWord5 m, UnsafeWord5 n) = map UnsafeWord5 $ range (m, n)
+  index (UnsafeWord5 m, UnsafeWord5 n) (UnsafeWord5 i) = index (m, n) i
+  inRange (m,n) i = m <= i && i <= n
+
+word5 :: Integral a => a -> Word5
+word5 x = UnsafeWord5 ((fromIntegral x) .&. 31)
+{-# INLINE word5 #-}
+{-# SPECIALIZE INLINE word5 :: Word8 -> Word5 #-}
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+{-# INLINE fromWord5 #-}
+{-# SPECIALIZE INLINE fromWord5 :: Word5 -> Word8 #-}
+
+charset :: Arr.Array Word5 Char
+charset = Arr.listArray (UnsafeWord5 0, UnsafeWord5 31) "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+charsetMap :: Char -> Maybe Word5
+charsetMap c | inRange (Arr.bounds inv) upperC = inv Arr.! upperC
+             | otherwise = Nothing
+  where
+    upperC = toUpper c
+    inv = Arr.listArray ('0', 'Z') (repeat Nothing) Arr.// (map swap (Arr.assocs charset))
+    swap (a, b) = (toUpper b, Just a)
+
+bech32Polymod :: [Word5] -> Word32
```

Comment:
One last comment.  I know I said to have `bech32Polymod` return a `Word32`, however now that I've studied the issue, if you think it is better, you could return `Word`.  My understanding of the Haskell report is that `Data.Word.Word` is guaranteed to be at least 30 bits (although this isn't stated quite as clearly as I would like).  Using `Int` as in your original code does work, but I feel that all bit-shifting work should be done on unsigned words to avoid the cognitive load of thinking about how sign-extension may or may not interact with the code.

Long story short, you can return `Word` if you think it is better.  I think both `Word` or `Word32` are fine.